### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by dantownsend** [piccolo-orm/piccolo#126] [Ideas]
Discussing the future direction of Piccolo.

**Comment by dantownsend** [comment]
Some good suggestions made by @heliumbrain https://github.com/piccolo-orm/piccolo/issues/10#issuecomment-871776533

I've started creating some GitHub project boards, for organising [enhancements](https://github.com/piccolo-orm/piccolo/projects/1) and [bugs](https://github.com/piccolo-orm/piccolo/projects/2).

I'll retire the [roadmap issue](https://github.com/piccolo-orm/piccolo/issues/10), and use this discussion thread instead. Any ideas listed there will be turned turned into issues, and will add them to the enhancements board.

**Comment by aminalaee** [comment]
@dantownsend Here some random ideas:

- Support for PK UUID ( I can pick that up)
- PostGIS support ?
- Random model builders for testing (there's usually a side-project like django model-bakery or blender) but that can be in project
-  Pydantic validation
- Abstract tables for inheritance

**Reply by dantownsend** [reply]
They are some great suggestions.

## UUID Primary Key

Using UUIDs for the primary key is an often requested feature. It might be possible to hack it just by doing this:

```python
from piccolo.table import Table
from piccolo.columns.column_types import UUID

class MyTable(Table):
    id = UUID(primary=True, key=True)
```

Or maybe make it configurable in the metaclass:

```python
class MyTable(Table, primary_uuid=True):
    ...
```
I haven't had time to play around with this much, as all of my own projects use auto incrementing integers for primary keys, but it's worth investigating.

It's also important it still works as expected with the Piccolo Admin.

## PostGIS support

There was some suggestion here about implementing PostGIS column types: https://github.com/piccolo-orm/piccolo/issues/55. In theory it shouldn't be too bad. It could be implemented as a third party library, but would also be cool to have it in the core Piccolo library, as some interesting UI could be made with it in the Piccolo Admin.

## Model builders

Yes, this is a nice idea. I'm a big fan of Django Dynamic Fixtures, which is similar. Something like this for Piccolo would be cool, perhaps based on [faker](https://faker.readthedocs.io/en/master/).

## Pydantic validation

Within Piccolo API, there's a function called `create_pydantic_model` ([docs](https://piccolo-api.readthedocs.io/en/latest/crud/serializers.html?highlight=create_pydantic#create-pydantic-model)), which creates Pydantic models from Piccolo tables. There has been some discussion about merging this into the core Piccolo library, or at the very least advertising it's existence better, as it's so important for FastAPI.

I don't want to tie Piccolo too tightly to Pydantic, so it's a difficult balancing act.

## Abstract tables for inheritance

Yes, this shouldn't be too hard to implement. The API could be as simple as:

```python
from piccolo.table import Table

class MyTable(Table, abstract=True):
    ...
```
And then those tables are ignored for the purposes of migrations (perhaps by adding a `get_table_classes` method to `AppConfig` which filters out abstract tables).

**Reply by aminalaee** [reply]
I played around the UUID primary key a bit.
Since we're creating table.id as PrimaryKey, `id = UUID(primary=True, key=True)` doesn't work. We add id by default as Integer or Serial to be more precise and overrides any other Primary Key.
I started working on a PR to add support for any column as Primary Key, I think I can get that out in a few days.
I'll also create an issue to discuss the approach.

**Reply by dantownsend** [reply]
Great, thanks!

**Reply by aminalaee** [reply]
@dantownsend What do you think about having some more managed fields/tables.
For example I've seen `GORM` in `Go` provides the ability to have a somewhat managed models where it can automatically add `created_at`, `updated_at`, even `deleted_at` for soft-deleting models.
I'm thinking something like `class MyTable(Table, managed=True)` can do that.

**Reply by dantownsend** [reply]
@aminalaee It's a nice idea. I wonder how descriptive `managed` is though - perhaps it could be called something like `auto_timestamped`?

**Reply by aminalaee** [reply]
Yeah, that's better.
